[quant][qat] Ensure fake_quant and observer can be disabled on scriptmodule

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44846 [quant] Add quant APIs to save/load observer state_dict
* **#44773 [quant][qat] Ensure fake_quant and observer can be disabled on scriptmodule**
* #44765 [quant][fx] Add node name as prefix to observer module name

Summary:
The model is created and prepared using fx APIs and then scripted for training.
In order to test QAT on scriptmodel we need to be able to disable/enable fake_quant
and observer modules on it.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qat_and_script

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23741354](https://our.internmc.facebook.com/intern/diff/D23741354)

supriyar committed:
[quant][qat] Ensure fake_quant and observer can be disabled on scriptmodule

Summary:
The model is created and prepared using fx APIs and then scripted for training.
In order to test QAT on scriptmodel we need to be able to disable/enable fake_quant
and observer modules on it.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qat_and_script

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fake_quantize.py:
```diff
@@ -156,17 +156,25 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
-    if type(mod) == FakeQuantize:
+    if type(mod) == FakeQuantize or\
+        isinstance(mod, torch.jit.RecursiveScriptModule)\
+            and mod.original_name == 'FakeQuantize':
```
is there a way to get the full qualified path?

supriyar replied:
You mean like `torch.quantization.FakeQuantize`? I wasn't able to find a way to do that.

jerryzh168 replied:
yeah, can you try to see if this works? https://github.com/pytorch/pytorch/blob/master/torch/_jit_internal.py#L736

jerryzh168 replied:
oh this one:https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/python/script_init.cpp.html#1046

you can probably get this by `m._c. qualified_name`

supriyar committed:
Update on "[quant][qat] Ensure fake_quant and observer can be disabled on scriptmodule"


Summary:
The model is created and prepared using fx APIs and then scripted for training.
In order to test QAT on scriptmodel we need to be able to disable/enable fake_quant
and observer modules on it.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qat_and_script

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][qat] Ensure fake_quant and observer can be disabled on scriptmodule"


Summary:
The model is created and prepared using fx APIs and then scripted for training.
In order to test QAT on scriptmodel we need to be able to disable/enable fake_quant
and observer modules on it.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qat_and_script

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/quantization/fake_quantize.py:
```diff
@@ -155,17 +155,25 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
                                                       reduce_range=True)
 def disable_fake_quant(mod):
-    if type(mod) == FakeQuantize:
+    if type(mod) == FakeQuantize or\
+        isinstance(mod, torch.jit.RecursiveScriptModule)\
+            and mod.original_name == 'FakeQuantize':
```
to clarify, why doesn't it work without this code?

supriyar replied:
Since it is scripted, the type will be `class 'torch.jit._script.RecursiveScriptModule` so the old code will return false.

supriyar committed:
Update on "[quant][qat] Ensure fake_quant and observer can be disabled on scriptmodule"


Summary:
The model is created and prepared using fx APIs and then scripted for training.
In order to test QAT on scriptmodel we need to be able to disable/enable fake_quant
and observer modules on it.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qat_and_script

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23741354](https://our.internmc.facebook.com/intern/diff/D23741354)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fake_quantize.py:
```diff
@@ -154,18 +155,30 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+        suffix = mod._c.qualified_name.split('.', 1)[1]
+        name = re.sub(r'\.___torch_mangle_\d+', '', suffix)
+        return name == 'torch.quantization.fake_quantize.FakeQuantize'
+    else:
+        return false
```
nit: this can be `return False`(without else), also it's `False` instead 'false` I think.

jerryzh168 reviewed (approved):
Looks good, thanks

jerryzh168 commented on torch/quantization/fake_quantize.py:
```diff
@@ -154,18 +155,30 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+    '''
+    if isinstance(mod, torch.jit.RecursiveScriptModule):
+        # qualified name looks like '__torch__.torch.quantization.fake_quantize.___torch_mangle_2.FakeQuantize'
+        suffix = mod._c.qualified_name.split('.', 1)[1]
```
I think it should be OK to include `__torch__`

supriyar committed:
Update on "[quant][qat] Ensure fake_quant and observer can be disabled on scriptmodule"


Summary:
The model is created and prepared using fx APIs and then scripted for training.
In order to test QAT on scriptmodel we need to be able to disable/enable fake_quant
and observer modules on it.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qat_and_script

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23741354](https://our.internmc.facebook.com/intern/diff/D23741354)

[ghstack-poisoned]
